### PR TITLE
Java: Apply deadcode guard to data flow nodes.

### PR DESCRIPTION
**Pull request by aschackmull**
Refactors the compile-time constant computation, such that we can reuse it for more general constant calculations.
Use this to define a trivial dead-code guard and apply it to data flow nodes to filter all flow through dead code.

It was necessary to reshuffle the dependencies between SSA, virtual dispatch, and data flow nodes a bit in order to be able to refer to SSA from data flow nodes. This means that field SSA uses a slightly lower precision virtual dispatch in the determination of uncertain updates of field-SSA variables. This is in line with what's also done in C#.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/Constants.qll [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    exists(ConditionalExpr ce, boolean condition |
+      ce = e and
+      condition = getBoolVal(ce.getCondition()) and
+      result = getBoolVal(ce.getBranchExpr(condition))
+    )
+    or
+    // If a `Variable` is final, its value is its initializer, if it exists.
+    exists(Variable v | e = v.getAnAccess() and v.isFinal() |
+      result = getBoolVal(v.getInitializer())
+    )
```
Comment (hvitved) [inline java/ql/lib/semmle/code/java/Constants.qll:98]: Should you consider adding SSA reads where you know the constant value of the SSA variable?

Comment (hvitved) [reply java/ql/lib/semmle/code/java/Constants.qll:98]: Same for `calculateIntValue`.

File: java/ql/lib/semmle/code/java/dataflow/RangeUtils.qll [hunk 2]
```diff
@@ -86,8 +87,29 @@ private predicate constantIntegerExpr(Expr e, int val) {
...
   )
+  or
+  CalcConstants::calculateIntValue(e) = val
+}
+
+pragma[nomagic]
+private predicate constantBooleanExpr(Expr e, boolean val) {
+  e.(CompileTimeConstantExpr).getBooleanValue() = val
+  or
+  exists(SsaExplicitUpdate v, Expr src |
```
Comment (hvitved) [inline java/ql/lib/semmle/code/java/dataflow/RangeUtils.qll:98]: I see that you include SSA steps here; why not add that inside the parameterized module?

**Review by hvitved — COMMENTED** [review]
LGTM, one question.

File: java/ql/lib/semmle/code/java/Constants.qll (hunk 1 above)
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/Constants.qll:98]: No, this file must include only exactly those steps that define compile-time-constants as per the Java Language Specification.

File: java/ql/lib/semmle/code/java/dataflow/RangeUtils.qll (hunk 2 above)
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/dataflow/RangeUtils.qll:98]: See above.

File: java/ql/lib/semmle/code/java/Constants.qll (hunk 1 above)
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/Constants.qll:98]: I've tried to specify this in the qldoc for the module.

Comment (hvitved) [reply java/ql/lib/semmle/code/java/Constants.qll:98]: I see that now, thanks.

**Comment by aschackmull** [comment]
The qltests caught a small corner-case: In ArrayIndexOutOfBounds.ql if we would previously catch a case based on a relative bound to an array length field access that happened to be a constant due to the array being defined in a final field and with a fixed length, then the additional constant propagation means that such a length access plus another constant is now found up front to be a constant, and thus the relative bound becomes superfluous, but the query didn't properly account for that case. This is now fixed. I've started an additional dca just for that query.

**Comment by aschackmull** [comment]
Looks like this causes ArrayIndexOutOfBounds.ql to catch a few more results, but it apparently also loses a few - that's worth an additional look before merging.

**Comment by zbazztian** [comment]
@starcke Thanks for doing this! Are you planning to add some flow test cases using this new feature? If not, I am happy to contribute them.

**Comment by zbazztian** [comment]
^
@starcke Apologies, I meant the other Anders, @aschackmull !

**Comment by aschackmull** [comment]
> Looks like this causes ArrayIndexOutOfBounds.ql to catch a few more results, but it apparently also loses a few - that's worth an additional look before merging.

One result lost to the very slight change in SSA precision for fields, and two results lost that turned out to actually be FPs.

**Review by hvitved — APPROVED** [review]

**Comment by aschackmull** [comment]
> Are you planning to add some flow test cases using this new feature? If not, I am happy to contribute them.

Feel free to add some as a follow-up. I wanted to merge this sooner rather than later, since it had been open for so long.

**Comment by zbazztian** [comment]
Makes sense. Thanks for implementing this, I will definitely give this a test this week.

**Pull request merged**